prov/efa/src/rxr/rxr_pkt_cmd.c

```diff
@@ -966,6 +966,130 @@ void rxr_pkt_proc_received(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
+ *        the address of old peer.
+ *     7. The packets from the old (destroyed) peer arrived.
+ *
+ * For the packets from destroyed peer, because they are from same GID + QPN,
+ * their pkt_entry->addr is same as new peer. By using connid in the header,
+ * we can identify these packets and mark them as "should be ignored"
+ *
+ * This function also insert address for REQ packets from a newly created peer
+ * (packets received on step 5 in the above workflow), and set
+ * pkt_entry->addr of these packets properly.
+ *
+ * @param[in]	ep		endpoint
+ * @param[in]	pkt_entry	received packet entry
+ * @return	a boolean value indicating whether this packet should be ignored
+ */
+static inline
+bool rxr_pkt_should_ignore_received(struct rxr_ep *ep,
+				    struct rxr_pkt_entry *pkt_entry)
+{
+	uint32_t *connid;
+	struct rdm_peer *peer;
+	struct efa_ep_addr *peer_addr; /* current peer address */
+	struct efa_ep_addr *sender_addr; /* peer address in packet header */
+
+	if (rxr_get_base_hdr(pkt_entry->pkt)->type >= RXR_REQ_PKT_BEGIN &&
+	    rxr_pkt_req_raw_addr(pkt_entry)) {
+		/* A REQ packet with raw address header can be the 1st packet
+		 * the endpoint received from a peer. Thus we need to be
+		 * careful here:
+		 *
+		 * the packet can be ignored only if its connid
+		 * matches previous qkey. In this case, we know for sure
+		 * this packet is from a destroyed peer.
+		 *
+		 * In all other case, we need to update AV with the new address
+		 * and process the packet.
+		 */
+		peer = rxr_ep_get_peer(ep, pkt_entry->addr);
+		peer_addr = rxr_peer_raw_addr(ep, pkt_entry->addr);
+
+		sender_addr = rxr_pkt_req_raw_addr(pkt_entry);
+		assert(sender_addr);
+		if (peer && sender_addr->qkey == peer->prev_qkey) {
+			FI_WARN(&rxr_prov, FI_LOG_CQ, "Ignoring a packet from a destroyed peer!\n"
+				"packet sender qkey: %d matches peer->prev_qkey: %d\n",
+				sender_addr->qkey, peer->prev_qkey);
+			return true;
+		}
+
+		if (!peer || sender_addr->qkey != peer_addr->qkey) {
+			/* If peer is NULL, pkt_entry->addr must be FI_ADDR_NOTAVAIL, which means
+			 * this packet is from a newly created peer and we have never received from this GID+QPN before.
+			 *
+			 * If peer is not NULL and qkey does not match, this packet is also from a newly created peer,
+			 * but the old peer with same GID+QPN has been destroyed.
+			 *
+			 * Either way, the new address need to be inserted to address vector, and
+			 * this packet cannot be ignored.
+			 */
+			pkt_entry->addr = rxr_pkt_insert_addr(ep, pkt_entry, sender_addr);
+		}
+
+		return false;
+	}
+
+	/*
+	 * Receiving a non-req packets or req packets without raw address header from a peer means the endpoint had prior
+	 * commuincation with the peer
+	 */
+	if (pkt_entry->addr == FI_ADDR_NOTAVAIL) {
+		/* If the endpoint had prior communication with the peer, but pkt_entry->addr is FI_ADDR_NOTAVAIL,
+		 * it means application called fi_av_remove() to remove the peer from address vector.
```

Comment:
Similar to above, we might want to keep this info around so we can double check that we were in fact communicating with this peer before. Otherwise we might hide bugs. We get to decide how to handle `fi_av_remove` so it's okay for us to hold onto those structures for this purpose.

Reply:
For this particular case, it is that we have removed an address from AV, then received a packet from that address. Thus the address is `FI_ADDR_NOTAVAIL`. So I am not sure where to store the address.